aws-rds-dbinstance/aws-rds-dbinstance.json

```diff
@@ -0,0 +1,431 @@
+          "minLength": 1,
+          "maxLength": 128
+        },
+        "Value": {
+          "type": "string",
+          "description": "The value for the tag. You can specify a value that is 1 to 255 Unicode characters in length and cannot be prefixed with aws:. You can use any of the following characters: the set of Unicode letters, digits, whitespace, _, ., /, =, +, and -. ",
+          "minLength": 1,
+          "maxLength": 256
+        }
+      },
+      "required": [
+        "Key",
+        "Value"
+      ]
+    }
+  },
+  "properties": {
+    "AllocatedStorage": {
+      "type": "integer",
+      "description": "The amount of storage (in gigabytes) to be initially allocated for the database instance."
+    },
+    "AllowMajorVersionUpgrade": {
+      "type": "boolean",
+      "description": "A value that indicates whether major version upgrades are allowed. Changing this parameter doesn't result in an outage and the change is asynchronously applied as soon as possible."
+    },
+    "AssociatedRoles": {
+      "type": "array",
+      "uniqueItems": true,
+      "items": {
+        "$ref": "#/definitions/DBInstanceRole"
+      },
+      "description": "The AWS Identity and Access Management (IAM) roles associated with the DB instance."
+    },
+    "AutoMinorVersionUpgrade": {
+      "type": "boolean",
+      "description": "A value that indicates whether minor engine upgrades are applied automatically to the DB instance during the maintenance window. By default, minor engine upgrades are applied automatically."
+    },
+    "AvailabilityZone": {
+      "type": "string",
+      "description": "The Availability Zone (AZ) where the database will be created. For information on AWS Regions and Availability Zones."
+    },
+    "BackupRetentionPeriod": {
+      "type": "integer",
+      "minimum": 0,
+      "maximum": 35,
+      "default": 1,
+      "description": "The number of days for which automated backups are retained. Setting this parameter to a positive number enables backups. Setting this parameter to 0 disables automated backups."
+    },
+    "CACertificateIdentifier": {
+      "type": "string",
+      "description": "The identifier of the CA certificate for this DB instance."
+    },
+    "CharacterSetName": {
+      "type": "string",
+      "description": "For supported engines, indicates that the DB instance should be associated with the specified character set."
+    },
+    "CopyTagsToSnapshot": {
+      "type": "boolean",
+      "description": "A value that indicates whether to copy tags from the DB instance to snapshots of the DB instance. By default, tags are not copied."
+    },
+    "DBClusterIdentifier": {
+      "type": "string",
+      "description": "The identifier of the DB cluster that the instance will belong to.",
+      "pattern": "^[a-zA-Z]{1}(?:-?[a-zA-Z0-9]){0,62}$"
```

Comment:
As it mentioned here:https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_CreateDBCluster.html under *DBInstanceIdentifier* parameter. The maximum length is 63 letters
This regex will accept string with length up to 126. For Example: this is a test case will pass this pattern

a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a-a

I would suggest you use this to avoid complex patterns and simplicity 
```
"pattern": "^[a-zA-Z](?:-?[a-zA-Z0-9])$",
"minLength": 1,
"maxLength": 63
```

Reply:
Amazing catch! Regarding using a pattern together with length limiters: the cfn tool warns about it. Idiomatically there is a chance of a conflict of these 3 constraints, so the best practice is to either specify min/max or a pattern.

Reply:
About the pattern itself: we use it in multiple places, e.g.:
* https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-rds/blob/master/aws-rds-dbcluster/aws-rds-dbcluster.json#L49
* https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-rds/blob/master/aws-rds-globalcluster/aws-rds-globalcluster.json#L26
I believe it is kept this way for the sake of simplicity (the no-dual hyphen means a look-ahead regex move with back-referencing).
I believe we should not treat this validation as the last border control to the resource definition. WS will do per-instance validation, in this case a CFN user will see an error once the request is sent to the backend. In other words, the CFN validation could be much more relaxed and not cover all possible cases (which might be pretty flexible and dynamic on the server side).